datalab/datalab_session/tasks.py

```diff
@@ -10,4 +16,19 @@ def execute_data_operation(data_operation_name: str, input_data: dict):
     if operation_class is None:
         raise NotImplementedError("Operation not implemented!")
     else:
-        operation_class(input_data).operate()
+        operation = operation_class(input_data)
+        cache_key = operation.generate_cache_key()
+
+        # check if we've done this operation already
+        if key_exists(operation.generate_cache_key()):
+            log.info(f'Operation {cache_key} cached')
+            
+            large_jpg_url = get_presigned_url(f'{cache_key}/{cache_key}-large.jpg')
```

Comment:
This doesn't seem like it will work for operations that generate more than one image. This should do something more generically - the output is cached so you should know what thumbnail files "should" exist from there, and then you could check each of those s3 urls and if they fail generate a new url and update the output. If you just set the presigned s3 url duration to 30 days and we delete the operation output at 30 days them you don't even need to worry about this here, it will just get re-generated when the operation has to run again.